app/code/Magento/CacheInvalidate/Observer/InvalidateVarnishObserver.php

```diff
@@ -57,10 +71,21 @@ public function execute(\Magento\Framework\Event\Observer $observer)
         if ($this->config->getType() == \Magento\PageCache\Model\Config::VARNISH && $this->config->isEnabled()) {
             $bareTags = $this->getTagResolver()->getTags($object);
 
+            $tagsBatchSize = 0;
             $tags = [];
-            $pattern = "((^|,)%s(,|$))";
+            $pattern = '((^|,)%s(,|$))';
             foreach ($bareTags as $tag) {
-                $tags[] = sprintf($pattern, $tag);
+                $formattedTag = sprintf($pattern, $tag);
+
+                // Send request if batch size is reached and add the implode with pipe to the computation
+                if ($tagsBatchSize + strlen($formattedTag) > $this->requestSize - count($tags) - 1) {
+                    $this->purgeCache->sendPurgeRequest(implode('|', array_unique($tags)));
```

Comment:
> This will allow customize this service for 3rd party developers in case they need some other behavior.

Do you have an example in mind?

As a reminder, the point of this PR is to fix the tags purging with Varnish because it does not work on big catalogs and I kept it as close as possible to what it was to avoid as much as possible any regression.

I am not against your idea, it is a nice enhancement but not at the cost of postponing this fix. This PR is a bugfix, your suggestion is an enhancement, and both should be on different roadmap.